# 🧠 Digit Classification using ANN (MNIST)
**Goal:** Train a basic Artificial Neural Network (ANN) to classify handwritten digits using the MNIST dataset and prepare it for integration with a React frontend.


In [2]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

## 📊 Load and Preprocess MNIST Dataset
- Normalize pixel values
- One-hot encode the labels

In [5]:
# Load MNIST
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalize to 0–1 range
x_train = x_train / 255.0
x_test = x_test / 255.0

# One-hot encode labels
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

## 🔧 Build the ANN
- Input: 28x28 image (flattened)
- Hidden Layer: Dense(128), ReLU
- Output Layer: Dense(10), Softmax

In [8]:
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.summary()

C:\Users\tharu\anaconda3\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

## 🏋️ Train the ANN
- Epochs: 10
- Batch Size: 32

In [11]:
history = model.fit(x_train, y_train_cat, epochs=10, batch_size=32, validation_data=(x_test, y_test_cat))


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.8754 - loss: 0.4371 - val_accuracy: 0.9574 - val_loss: 0.1454
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9647 - loss: 0.1240 - val_accuracy: 0.9669 - val_loss: 0.1105
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9759 - loss: 0.0803 - val_accuracy: 0.9723 - val_loss: 0.0873
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9813 - loss: 0.0586 - val_accuracy: 0.9759 - val_loss: 0.0784
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9866 - loss: 0.0435 - val_accuracy: 0.9753 - val_loss: 0.0764
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9894 - loss: 0.0341 - val_accuracy: 0.9769 - val_loss: 0.0770
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.9913 - loss: 0.0286 - val_accuracy: 0.9782 - val_loss: 0.0739
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9936 - loss: 0.02

## ✅ Evaluate the Model & Save
- Check test accuracy
- Save model to use in Flask backend

In [14]:
test_loss, test_acc = model.evaluate(x_test, y_test_cat)
print(f'Test accuracy: {test_acc:.2f}')

# Save model
model.save("ann_model.h5")

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9735 - loss: 0.0914


Test accuracy: 0.98


# 🧠 Digit Classification with ANN + React UI

## 📌 Project Overview

This project demonstrates a complete AI application pipeline using a basic **Artificial Neural Network (ANN)** to classify handwritten digits from the **MNIST dataset**, with a frontend built in **React** where users can draw digits and get real-time predictions.

- **Model:** ANN (Input → Hidden → Output)
- **Dataset:** MNIST (70,000 grayscale 28x28 digit images)
- **Frontend:** React.js (Canvas + Prediction Display)
- **Backend:** Flask (Python API to serve the trained model)

---

## 🔧 Tech Stack

| Layer      | Tools Used                       |
|------------|----------------------------------|
| Model      | TensorFlow / Keras               |
| API Server | Flask, NumPy, Pillow             |
| Frontend   | React, Axios, HTML5 Canvas       |
| Data       | MNIST Dataset (from Keras)       |

---

## 📊 Model Architecture (ANN)

- Input Layer: Flattened 28x28 image (784 nodes)
- Hidden Layer: Dense(128), ReLU activation
- Output Layer: Dense(10), Softmax activation

```python
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])